### PR TITLE
feat: implement expect(locator).toBeInViewport()

### LINKED ISSUE #8740: [Feature] add expect(locator).isInViewport() / isIntersectingViewport
Now we have to use this method, the implementation of which is taken from puppeteer. It would be more convenient if it was in the playwright itself
```ts
  /**
   * @returns {!Promise<boolean>}
   * @summary Проверка, что нужный элемент находится в текущей области видимости(viewPort)
   */
  isIntersectingViewport(selector: string): Promise<boolean> {
    return this.page.$eval(selector, async element => {
      const visibleRatio: number = await new Promise(resolve => {
        const observer = new IntersectionObserver(entries => {
          resolve(entries[0].intersectionRatio);
          observer.disconnect();
        });
        observer.observe(element);
        // Firefox doesn't call IntersectionObserver callback unless
        // there are rafs.
        requestAnimationFrame(() => {});
      });
      return visibleRatio > 0;
    });
  }
```

Context: https://playwright.slack.com/archives/CSUHZPVLM/p1631008178339400

### PR DESCRIPTION
**Pull request by aslushnikov**
The method accepts a `ratio` option to assert the ratio
of the element in viewport. `ratio` defaults to `0`.

NOTE: this reverts commit d950f5b6ee3fee4b825831983d5af5b197bda769 and
adds `ratio` option support + does the rename.

Fixes #8740

### CODE REVIEW

File: docs/src/api/class-locatorassertions.md [hunk 1]
```diff
@@ -733,6 +733,27 @@ await Expect(locator).ToBeHiddenAsync();
 ### option: LocatorAssertions.toBeHidden.timeout = %%-csharp-java-python-assertions-timeout-%%
 * since: v1.18
 
+## async method: LocatorAssertions.toBeInViewport
+* since: v1.31
+* langs: js
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:738]: lets add it to the ports?

File: docs/src/api/class-locatorassertions.md [hunk 2]
```diff
@@ -733,6 +733,27 @@ await Expect(locator).ToBeHiddenAsync();
...
+
+Ensures the [Locator] points to an element that intersects viewport, according to the [intersection observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API).
+
+**Usage**
+
+```js
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:744]: lets add here snippets for the language ports

File: docs/src/api/class-locatorassertions.md [hunk 3]
```diff
@@ -733,6 +733,27 @@ await Expect(locator).ToBeHiddenAsync();
...
+const locator = page.locator('button.submit');
+await expect(locator).toBeInViewport();
+```
+
+### option: LocatorAssertions.toBeInViewport.ratio
+
+The minimal ratio of the element to be visible in viewport. Defaults to [`Number.MIN_VALUE`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MIN_VALUE).
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:751]: This makes it hard for the ports, I'd vote for something simpler like 0-1.

**Review by dgozman — APPROVED** [review]

File: docs/src/api/class-locatorassertions.md (hunk 3 above)
Comment (aslushnikov) [reply docs/src/api/class-locatorassertions.md]: sure. Switched to strict comparison; this defaults to `0` now.

File: docs/src/api/class-locatorassertions.md [hunk 4]
```diff
@@ -679,6 +679,80 @@ await Expect(locator).ToBeHiddenAsync();
...
+```
+
+```csharp
+var locator = Page.Locator("button.submit");
+// Make sure at least some part of element intersects viewport.
+await Expect(locator).ToBeInViewport();
+// Make sure element is fully outside of viewport.
+await Expect(locator).Not.ToBeInViewport();
+// Make sure strictly more than half of the element intersects viewport.
+await Expect(locator).ToBeInViewport(new() { Ratio = 0.5 });
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:718]:
```suggestion
await Expect(locator).ToBeInViewportAsync(new() { Ratio = 0.5 });
```

File: docs/src/api/class-locatorassertions.md [hunk 5]
```diff
@@ -679,6 +679,80 @@ await Expect(locator).ToBeHiddenAsync();
...
+// Make sure strictly more than half of the element intersects viewport.
+assertThat(locator).isInViewport(new LocatorAssertions.IsInViewportOptions().setRatio(0.5));
+```
+
+```csharp
+var locator = Page.Locator("button.submit");
+// Make sure at least some part of element intersects viewport.
+await Expect(locator).ToBeInViewport();
+// Make sure element is fully outside of viewport.
+await Expect(locator).Not.ToBeInViewport();
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:716]:
```suggestion
await Expect(locator).Not.ToBeInViewportAsync();
```

File: docs/src/api/class-locatorassertions.md [hunk 6]
```diff
@@ -679,6 +679,80 @@ await Expect(locator).ToBeHiddenAsync();
...
+// Make sure element is fully outside of viewport.
+assertThat(locator).not().isInViewport();
+// Make sure strictly more than half of the element intersects viewport.
+assertThat(locator).isInViewport(new LocatorAssertions.IsInViewportOptions().setRatio(0.5));
+```
+
+```csharp
+var locator = Page.Locator("button.submit");
+// Make sure at least some part of element intersects viewport.
+await Expect(locator).ToBeInViewport();
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:714]:
```suggestion
await Expect(locator).ToBeInViewportAsync();
```

File: docs/src/api/class-locatorassertions.md [hunk 7]
```diff
@@ -679,6 +679,80 @@ await Expect(locator).ToBeHiddenAsync();
...
+
+
+### option: LocatorAssertions.toBeInViewport.ratio
+* since: v1.31
+- `ratio` <[float]>
+
+The minimal ratio of the element to intersect viewport. Element's ratio should be strictly greater than
+this number. Defaults to `0`.
+
+### option: LocatorAssertions.toBeInViewport.timeout = %%-js-assertions-timeout-%%
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:753]: this is only the js timeout, what about the timeout for the other langs?

File: docs/src/api/class-locatorassertions.md [hunk 8]
```diff
@@ -679,6 +679,80 @@ await Expect(locator).ToBeHiddenAsync();
...
+// Make sure element is fully outside of viewport.
+await Expect(locator).Not.ToBeInViewport();
+// Make sure strictly more than half of the element intersects viewport.
+await Expect(locator).ToBeInViewport(new() { Ratio = 0.5 });
+```
+
+```python async
+from playwright.async_api import expect
+
+locator = page.locator('button.submit')
```
Comment (mxschmitt) [inline docs/src/api/class-locatorassertions.md:724]: double quote

**Pull request merged**